frameworks/Haskell/servant/mysql-haskell/src/Main.hs

```diff
@@ -0,0 +1,256 @@
+
+singleDb :: DbPool -> GenIO -> Handler World
+singleDb pool gen = do
+  v <- liftIO $ uniformR (1, 10000) gen
+  r <- liftIO $ Pool.withResource pool $ \conn -> do
+    (_, rowsS) <- MySQL.query conn "SELECT * FROM World WHERE id = ?" [intValEnc v]
+    extractWorld rowsS
+  case r of
+    Left e -> throwError err500 { errBody = LBS.fromStrict $ TextEnc.encodeUtf8 e }
+    Right world -> return world
+{-# INLINE singleDb #-}
+
+-- * Test 3: Multiple database query
+
+multipleDb :: DbPool -> GenIO -> Maybe Count -> Handler [World]
+multipleDb pool gen mcount = do
+  res <- liftIO $ Pool.withResource pool $ \conn -> do
+    sId <- MySQL.prepareStmt conn "SELECT * FROM World WHERE id = ?"
+    res <- replicateM (getCount mcount) $ do
+      v <- uniformR (1, 10000) gen
+      (_, rowsS) <- MySQL.queryStmt conn sId [intValEnc v]
+      extractWorld rowsS
+    MySQL.closeStmt conn sId
+    return res
+  let (errs, oks) = partitionEithers res
+  case errs of
+    [] -> return oks
+    e -> throwError err500 { errBody = LBS.fromStrict . TextEnc.encodeUtf8 . Text.pack . show $ e }
+{-# INLINE multipleDb #-}
+
+
+-- * Test 4: Fortunes
+
+decodeFortune :: DbRow -> Either Text Fortune
+decodeFortune [] = Left "MarshalError: Expected 2 columns for Fortune, found 0"
+decodeFortune (_:[]) = Left "MarshalError: Expected 2 columns for Fortune, found 1"
+decodeFortune (c1:c2:_) = Fortune <$> intValDec c1 <*> textValDec c2
+{-# INLINE decodeFortune #-}
+
+selectFortunes :: MySQL.MySQLConn -> IO (Either [Text] [Fortune])
+selectFortunes conn = do
+  (_, rowsS) <- MySQL.query_ conn "SELECT * FROM Fortune"
+  rows <- Streams.toList rowsS
+  let eFortunes = fmap decodeFortune rows
+  let (err, oks) = partitionEithers eFortunes
+  return $ case err of
+    [] -> pure oks
+    _ -> Left err
+{-# INLINE selectFortunes #-}
+
+fortunes :: DbPool -> Handler (Lucid.Html ())
+fortunes pool = do
+  r <- liftIO $ Pool.withResource pool selectFortunes
+  case r of
+    Left e -> throwError err500 { errBody = LBS.fromStrict . TextEnc.encodeUtf8 . Text.pack . show $ e }
+    Right fs -> return $ do
+      let new = Fortune 0 "Additional fortune added at request time."
+      Lucid.doctypehtml_ $ do
+        Lucid.head_ $ Lucid.title_ "Fortunes"
+        Lucid.body_ $ do
+          Lucid.table_ $ do
+            Lucid.tr_ $ do
+              Lucid.th_ "id"
+              Lucid.th_ "message"
+            mapM_ (\f -> Lucid.tr_ $ do
+              Lucid.td_ (Lucid.toHtml . show $ fId f)
+              Lucid.td_ (Lucid.toHtml $ fMessage f)) (sortOn fMessage (new : fs))
+{-# INLINE fortunes #-}
+
+-- * Test 5: Updates
+
+updates :: DbPool -> GenIO -> Maybe Count -> Handler [World]
+updates pool gen mcount = do
+  res <- liftIO $ Pool.withResource pool $ \conn -> do
+    sIdGet <- MySQL.prepareStmt conn "SELECT * FROM World WHERE id = ?"
+    sIdPut <- MySQL.prepareStmt conn "UPDATE World SET randomNumber = ? WHERE id = ?"
+    res <- replicateM (getCount mcount) $ do
```

Comment:
I forget - is against the rules/idea to use `executeMany`?

Reply:
> use of IN clauses or similar means to consolidate multiple queries into one operation is not permitted

I think multiple database "operations"/calls is absolutely required.

Reply:
This's probably the best we can get from mysql-haskell, how much is the slowdown comparing to hasql?

Reply:
Raw results are available in the main post: 
https://github.com/TechEmpower/FrameworkBenchmarks/pull/4550#issue-261807511

Here be the slow(er) stats:

Multiple queries

lib|20-queries (bar)|Latency(ms)
---|---|---
hasql|375|749.4
mysql-haskell|375|1300.0

Updates

lib|20-queries (bar)|Latency(ms)
---|---|---
hasql|174|1300.0
mysql-haskell|90|4700.0

Reply:
it's curious because single queries are faster. I thought it would be faster all-round because pure Haskell drivers do not have the ffi/blocking issue.

on a separate note: would std-io (when it's stable and ready) help someday?